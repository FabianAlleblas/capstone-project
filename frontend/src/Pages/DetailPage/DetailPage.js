import { useHistory, useLocation } from 'react-router-dom'
import styled from 'styled-components/macro'
import Header from '../../Components/Header/Header'
import PlantDetailCard from '../../Components/Plantdetailcard/PlantDetailCard'

export default function DetailPage({ plantList }) {
  const history = useHistory()
  const query = useQuery()
  const plantId = parseInt(query.get('id'))
  const plant = plantList.find((plant) => plant.id === plantId)

  return (
    <>
      <FixedHeader
        isDetailPage
        onClickArrow={openPlantList}
        onClickEdit={openEditPage}
      >
        {plant?.name}
      </FixedHeader>
      <CardContainer>
        <PlantDetailCard species={plant?.species} info={plant?.info} />
      </CardContainer>
    </>
  )

  function useQuery() {
    return new URLSearchParams(useLocation().search)
  }

  function openPlantList() {
    history.push('/')
  }

  function openEditPage() {
    history.push(`/edit?id=${plant.id}`)
  }
}

const FixedHeader = styled(Header)`
  left: 0;
  position: fixed;
  top: 0;
`

const CardContainer = styled.main`
  padding: 100px 20px 0;
`
