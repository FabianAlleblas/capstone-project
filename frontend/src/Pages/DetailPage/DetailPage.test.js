import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import 'jest-styled-components'
import { MemoryRouter } from 'react-router-dom'
import defaultPlantImage from '../../assets/plant-images/default_plant.jpg'
import DetailPage from './DetailPage'

const plantList = [
  {
    id: 2,
    name: 'Carl',
    species: 'Monstera deliciosa',
    specialInfo: 'variegated',
    daysLeft: 5,
    weeksLeft: 2,
    waterInterval: 10,
    fertilizerInterval: 4,
  },
]

const mockHistoryPush = jest.fn()

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}))

const resetTimerMock = jest.fn()
const updateCareIntervalMock = jest.fn()

describe('DetailPage', () => {
  it('renders correctly', () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/plant?id=2']}>
        <DetailPage
          plantList={plantList}
          resetTimer={resetTimerMock}
          updateCareInterval={updateCareIntervalMock}
        />
      </MemoryRouter>
    )
    expect(container).toMatchSnapshot()
  })

  it('shows the correct data', () => {
    const { getByText, getByTestId } = render(
      <MemoryRouter initialEntries={['/plant?id=2']}>
        <DetailPage
          plantList={plantList}
          resetTimer={resetTimerMock}
          updateCareInterval={updateCareIntervalMock}
        />
      </MemoryRouter>
    )

    const WaterBar = getByTestId('water-indicator')
    const FertilizerBar = getByTestId('fertilizer-indicator')

    expect(WaterBar).toHaveStyle('width: 50%;')
    expect(FertilizerBar).toHaveStyle('width: 50%')

    expect(getByText(/carl/i)).toBeInTheDocument()

    expect(getByTestId('image-frame')).toHaveStyle(
      `background-image: url(${defaultPlantImage})`
    )
  })

  it('calls history push by clicking the back button', () => {
    const { getByRole } = render(
      <MemoryRouter initialEntries={['/plant?id=2']}>
        <DetailPage
          plantList={plantList}
          resetTimer={resetTimerMock}
          updateCareInterval={updateCareIntervalMock}
        />
      </MemoryRouter>
    )

    user.click(getByRole('button', { name: /arrowicon/i }))
    expect(mockHistoryPush).toHaveBeenCalled()
  })

  it('calls history push with the correct data by clicking the edit button', () => {
    const { getByRole } = render(
      <MemoryRouter initialEntries={['/plant?id=2']}>
        <DetailPage
          plantList={plantList}
          resetTimer={resetTimerMock}
          updateCareInterval={updateCareIntervalMock}
        />
      </MemoryRouter>
    )

    user.click(getByRole('button', { name: /editicon/i }))
    expect(mockHistoryPush).toHaveBeenCalledWith('/edit-plant?id=2')
  })

  it('renders the error modal without a correct id', () => {
    const { getByText } = render(
      <MemoryRouter initialEntries={['/plant?id=3']}>
        <DetailPage
          plantList={plantList}
          resetTimer={resetTimerMock}
          updateCareInterval={updateCareIntervalMock}
        />
      </MemoryRouter>
    )

    expect(getByText(/plant not found/i)).toBeInTheDocument()
  })
})
