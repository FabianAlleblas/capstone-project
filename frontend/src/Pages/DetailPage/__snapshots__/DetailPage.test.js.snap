// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailPage renders correctly 1`] = `
.c22 {
  background-color: var(--primary-light);
  border-radius: 22px;
  border: none;
  box-shadow: 0 0 4px var(--shadow);
  color: var(--primary-dark);
  font-family: 'Josefin sans';
  font-size: 1.25rem;
  font-weight: 700;
  padding: 12px 16px 8px;
  width: 100%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
  z-index: 10;
}

.c2 {
  -webkit-align-items: baseline;
  -webkit-box-align: baseline;
  -ms-flex-align: baseline;
  align-items: baseline;
  background: var(--primary-dark);
  border-radius: 0 0 50px 0;
  box-shadow: 0 0 4px var(--shadow);
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 20px 30px 13px 20px;
}

.c5 {
  color: var(--primary-light);
  font-size: 1.75rem;
  font-weight: 700;
  line-height: 1;
}

.c4 {
  margin-right: 12px;
}

.c6 {
  margin-right: 12px;
}

.c3 {
  background-color: transparent;
  border: none;
}

.c12 {
  display: grid;
  gap: 20px;
  padding: 10px 0 0;
  width: 100%;
}

.c13 {
  display: grid;
  gap: 10px;
}

.c14 {
  background-color: var(--bar-bg-color);
  border-radius: 2px;
  height: 4px;
  position: relative;
  width: 100%;
}

.c15 {
  border-radius: 4px;
  height: 8px;
  min-width: 10px;
  position: absolute;
  top: -2px;
}

.c16 {
  background-color: var(--bar-water-color);
  width: 50%;
}

.c20 {
  background-color: var(--bar-fertilizer-color);
  width: 50%;
}

.c17 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c18 {
  color: var(--secondary-plant-font-color);
  font-size: 1rem;
}

.c19 {
  background-color: transparent;
  border: none;
}

.c8 {
  background-image: url(default_plant.jpg);
  background-position: center;
  background-size: cover,contain;
  border-radius: 50%;
  height: 200px;
  width: 200px;
}

.c9 {
  line-height: 1.2;
  text-align: center;
}

.c10 {
  color: var(--primary-plant-font-color);
  font-size: 1.4rem;
  font-weight: 700;
}

.c11 {
  color: var(--secondary-plant-font-color);
  font-size: 1.4rem;
  font-weight: 400;
}

.c1 {
  left: 0;
  position: fixed;
  top: 0;
}

.c7 {
  display: grid;
  gap: 30px;
  padding: 100px 30px 30px;
  place-items: center;
}

.c21 {
  display: grid;
  gap: 20px;
  grid-template-columns: 1fr 1fr;
  padding: 20px 0 0;
  width: 100%;
}

.c23 {
  background-color: var(--primary-dark);
  color: var(--primary-light);
}

<div>
  <header
    class="c0 c1"
  >
    <div
      class="c2"
      data-testid="heading-container"
    >
      <button
        class="c3"
      >
        <svg
          class="c4"
        >
          arrowicon.svg
        </svg>
      </button>
      <h1
        class="c5"
      >
        Carl
      </h1>
    </div>
    <button
      class="c3"
    >
      <svg
        class="c6"
      >
        editicon.svg
      </svg>
    </button>
  </header>
  <section
    class="c7"
  >
    <div
      class="c8"
      data-testid="image-frame"
      src="default_plant.jpg"
    />
    <div
      class="c9"
    >
      <h2
        class="c10"
      >
        Monstera deliciosa
      </h2>
      <h3
        class="c11"
      >
        variegated
      </h3>
    </div>
    <div
      class="c12"
    >
      <div
        class="c13"
      >
        <div
          class="c14"
        >
          <div
            class="c15 c16"
            data-testid="water-indicator"
          />
        </div>
        <div
          class="c17"
        >
          <p
            class="c18"
          >
            Water 
            5
            /
            10
             days left
          </p>
          <button
            class="c19"
            data-testid="water-button"
            name="water"
          >
            <svg>
              settingsicon.svg
            </svg>
          </button>
        </div>
      </div>
      <div
        class="c13"
      >
        <div
          class="c14"
        >
          <div
            class="c15 c20"
            data-testid="fertilizer-indicator"
          />
        </div>
        <div
          class="c17"
        >
          <p
            class="c18"
          >
            Fertilizer 
            2
            /
            4
             weeks left
          </p>
          <button
            class="c19"
            data-testid="fertilizer-button"
            name="fertilizer"
          >
            <svg>
              settingsicon.svg
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="c21"
    >
      <button
        class="c22 c23"
      >
        Watered
      </button>
      <button
        class="c22 c23"
      >
        Fertilized
      </button>
    </div>
  </section>
</div>
`;
